cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,858 @@
+                 - 1./3. * trace(material_model_inputs.strain_rate[q]) * unit_symmetric_tensor<dim>()
+                 :
+                 material_model_inputs.strain_rate[q]);
+                
+                heating_model_outputs.heating_source_terms[q] = stress * compressible_strain_rate;
+                
+                // If dislocation viscosities and boundary area work fractions are provided, reduce the
+                // overall heating by this amount (which is assumed to increase surface energy)
+                if (disl_viscosities_out != 0)
+                {
+                    heating_model_outputs.heating_source_terms[q] *= 1 - disl_viscosities_out->boundary_area_change_work_fractions[q] *
+                    material_model_outputs.viscosities[q] / disl_viscosities_out->dislocation_viscosities[q];
+                }
+                
+                heating_model_outputs.lhs_latent_heat_terms[q] = 0.0;
+            }
+        }
+        
+        template <int dim>
+        void
+        ShearHeatingAnisotropicViscosity<dim>::
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &material_model_outputs) const
+        {
+            const unsigned int n_points = material_model_outputs.viscosities.size();
+            
+            if (material_model_outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >() == nullptr)
+            {
+                material_model_outputs.additional_outputs.push_back(
+                                                                    std::make_shared<MaterialModel::AnisotropicViscosity<dim>> (n_points));
+            }
+            
+            this->get_material_model().create_additional_named_outputs(material_model_outputs);
+        }
+    }
+    
+    namespace MaterialModel
+    {
+        using namespace dealii;
+        
+        template <int dim>
+        class Anisotropic : public MaterialModel::Simple<dim>
+        {
+        public:
+            virtual void initialize();
+            
+            virtual void evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+                                  MaterialModel::MaterialModelOutputs<dim> &out) const;
+            static void declare_parameters(ParameterHandler &prm);
+            virtual void parse_parameters(ParameterHandler &prm);
+            
+            /**
+             * Return true if the compressibility() function returns something that
+             * is not zero.
+             */
+            virtual bool
+            is_compressible () const;
+            
+        private:
+            void set_assemblers(const SimulatorAccess<dim> &,
+                                Assemblers::Manager<dim> &assemblers) const;
+            
+            // Constitutive tensor
+            SymmetricTensor<4,dim> C;
+        };
+        template <int dim>
+        class AV : public MaterialModel::Simple<dim>

```

Comment:
Can you add comments in front of this class and the previous one that explain what these classes do?